cell_classification/model_builder.py

```diff
@@ -63,21 +63,42 @@ def prep_data(self):
             ]
 
         # split into train, validation and test
-        self.validation_datasets = [
-            dataset.take(num_validation) for dataset, num_validation in zip(
+        if "data_splits" in self.params.keys():
+            data_splits = []
+            for fpath in self.params["data_splits"]:
+                with open(fpath, "r") as f:
+                    data_splits.append(json.load(f))
+            self.validation_datasets = [
+                self.fov_filter(dataset, data_split["validation"]) for dataset, data_split in zip(
+                    datasets, data_splits
+                )
+            ]
+            self.test_datasets = [
+                self.fov_filter(dataset, data_split["test"]) for dataset, data_split in zip(
+                    datasets, data_splits
+                )
+            ]
+            self.train_datasets = [
+                self.fov_filter(dataset, data_split["train"]) for dataset, data_split in zip(
+                    datasets, data_splits
+                )
+            ]
+        else:
+            self.validation_datasets = [
+                dataset.take(num_validation) for dataset, num_validation in zip(
```

Comment:
Where does the randomization happen? I see that the JSON files included properly randomized splits, so it's happening somewhere, but this just takes the first `x` entries, not a random `x`

Reply:
Yep, shuffling happens later. I did the train/val/test splits without shuffling, so that I have disjoint sets of FOVs in the dataset subsets. Shuffling the tiles before would mix FOVs in the subsets which I wanted to avoid.

Reply:
Right, but how are the FOVs themselves shuffled? i.e. not in alphabetical order?

Reply:
If you don't provide `data_splits` FOVs are in alphabetical order and the train/val/test split via `dataset.take()` will put them into subsets like this `[a, b], [c, d], [e,f,g,h,...,z]`. If you provide `data_splits`, it is assumed that FOVs in `data_splits` are already shuffled. After datasets are split in one way or the other, the tiles in the train dataset are shuffled.

Reply:
Okay, so in order to generate the TONIC `data_splits`, which look like `[[TMA1, TMA9], [[TMA22, TMA3], [TMA18, TMA14]]`, you had to provide the datasplits? The existing logic, if supplied no data_splits, would have instead done `[[TMA1, TMA2], [TMA3, TMA4], [...TMA18, TMA19?]`